src/world/Units/Unit.cpp

```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
     return false;
 }
 
+bool Unit::hasAuraWithAuraEffect(AuraEffect type) const
+{
+    for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
+    {
+        if (m_auras[i] == nullptr)
+            continue;
+        if (m_auras[i]->GetSpellInfo()->HasEffectApplyAuraName(type))
+            return true;
+    }
+    return false;
+}
+
+bool Unit::hasAuraState(AuraState state, SpellInfo* spellInfo, Unit* caster) const
+{
+    if (caster != nullptr && spellInfo != nullptr && caster->hasAuraWithAuraEffect(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+    {
+        for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
+        {
+            if (caster->m_auras[i] == nullptr)
+                continue;
+            if (!caster->m_auras[i]->GetSpellInfo()->HasEffectApplyAuraName(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+                continue;
+            if (caster->m_auras[i]->GetSpellInfo()->isAffectingSpell(spellInfo))
+                return true;
+        }
+    }
+    return unitData()->aura_state & (1 << (state - 1));
+}
+
+void Unit::addAuraState(AuraState state)
+{
+    if (!(unitData()->aura_state & (1 << (state - 1))))
+    {
+        write(unitData()->aura_state, unitData()->aura_state | uint32_t(1 << (state - 1)));
```

Comment:
Please isolate the write call into its own function and then call that, instead of mixing it with other logic.

Reply:
We will run into some naming issues here... addAuraState should be used to isolate the write function (unitData()).
This function does not simply add the aurastates.... it does also controle (checks) if the spell can be casted....
I would suggest to rename this function to addAuraStateAndCast so the write(unitData()) part can be isolated to addAuraState.

Reply:
@Appled I agree with Zyres' suggestion, assuming you're able to perform method rename refactoring.

If not, let us know and one of us will do it.